### PR TITLE
589 derive vars dtm ignore sec

### PR DESCRIPTION
**Pull request by bms63**

### CODE REVIEW

**Comment by thomas-neitmann** [comment]
The `R CMD check` workflow fails because the documentation is out of sync with the code. Please run `devtools::document()` and comit the chanend `.Rd` file in the `man` folder. There are also a couple fo lintr annotation that have to be addressed.

**Comment by bms63** [comment]
Hi all,

I'm having trouble with RMD Check.    When I `Run Tests` within the `test-derive_vars_dtm.R` file everything goes smoothly.  

image

But when I run the RMD Check it crashes.  Maybe my logic is bad within the functions that I updated.  Is it possible for you all to look over and provide any advice @thomas-neitmann @bundfussr @kabis-ops ?

## The First Error is within the example - which works fine on its own

image

## The Second Error 

Is when I'm invoking my ignore_seconds_flag = TRUE.


I'm guessing the issue lies within my logic and using two booleans?? https://github.com/r-lib/devtools/issues/2032

**Comment by kabis-ops** [comment]
Hi @bms63 - this seems to be due to `compute_tmf` and the condition on row 455 (see https://github.com/r-lib/testthat/issues/1176). If you change it to 
```
 if (ignore_seconds_flag)  {
    if (n_chr >= 17)
      {abort("Seconds detected in data while ignore_seconds_flag is invoked")
    }
    else {
      case_when(
        (!is_na & n_chr >= 19 & valid_dtc) | is_na | !valid_dtc ~ NA_character_,
        n_chr == 13 ~ "M",
        n_chr == 10 | (n_chr > 0 & n_chr < 10) ~ "H"
      )}
   }
  else {
    ...
}
```
this should pass the CMD check.

File: R/derive_date_vars.R [hunk 1]
```diff
@@ -724,9 +777,15 @@ derive_vars_dtm <- function(dataset,
     # add --TMF variable
     tmf <- paste0(new_vars_prefix, "TMF")
     warn_if_vars_exist(dataset, tmf)
+
     dataset <- dataset %>%
-      mutate(!!sym(tmf) := compute_tmf(dtc = !!dtc, dtm = !!sym(dtm)))
+      mutate(!!sym(tmf) := compute_tmf(
+        dtc = !!dtc,
+        dtm = !!sym(dtm),
+        ignore_seconds_flag = !!ignore_seconds_flag))
```
Comment (bundfussr) [inline R/derive_date_vars.R:785]: I think `!!` should be removed.

**Review by bundfussr — CHANGES_REQUESTED** [review]
Please update the Changelog (NEWS.md).

**Comment by bms63** [comment]
@kabis-ops  Thank you for the suggestion but it still doesn't seem to be working once passed up to RMD Checks.  I looked through other code and can this being done - either two conditions in an if statement or a nested if statement like you suggested.  

Feel like there is some other issue that I'm not seeing that is causing the problem.  Maybe the case_when function??

**Comment by bms63** [comment]
or how the compute_tmf is being used downstream??

**Comment by bundfussr** [comment]
Just a shot in the dark: You could try to remove `!!` in line 786. Maybe that's confusing dplyr.

**Comment by bms63** [comment]
No luck there!  Very strange error with the if/else statements as both nested if/else statements as well as if statements with multiple statements are used throughout this script.

I'm starting to think it has to do with how compute_tmf is used in the mutate() function later on in the script.

**Comment by bundfussr** [comment]
Maybe `if (n_chr >= 17)` is causing the issue because it results in a logical vector. You could try `if (any(n_chr >= 17))`.

**Comment by bms63** [comment]
Okay - lord have mercy.  Now we can actually start the review.  Many apologies for the mess and thank you for all the support.

File: R/derive_date_vars.R [hunk 2]
```diff
@@ -415,6 +418,15 @@ compute_dtf <- function(dtc, dt) {
 #'
 #'   A datetime object is expected.
 #'
+#' @param ignore_seconds_flag  ADaM IG states that given SDTM (`'--DTC'`) variable,
+#' if only hours and minutes are ever collected, and seconds are imputed in
+#' (`'--DTM'`) as 00, then it is not necessary to set (`'--TMF'`) to `'S'`.  A user can set this
+#' to `TRUE` so the  `'S'` Flag is dropped from (`'--TMF'`).
```
Comment (thomas-neitmann) [inline R/derive_date_vars.R:424]: There's one too much space in front of `'S'`.

File: R/derive_date_vars.R [hunk 3]
```diff
@@ -427,20 +439,35 @@ compute_dtf <- function(dtc, dt) {
 #' compute_tmf(dtc = "2019-07-18T15:25", dtm = as.POSIXct("2019-07-18T15:25:00"))
 #' compute_tmf(dtc = "2019-07-18T15", dtm = as.POSIXct("2019-07-18T15:25:00"))
 #' compute_tmf(dtc = "2019-07-18", dtm = as.POSIXct("2019-07-18"))
-compute_tmf <- function(dtc, dtm) {
+compute_tmf <- function(dtc,
+                        dtm,
+                        ignore_seconds_flag = FALSE) {
+
   assert_that(is.character(dtc), is_date(dtm))
```
Comment (thomas-neitmann) [inline R/derive_date_vars.R:446]: Can you use `assert:character_vector()` instead and maybe add a new `assert_date_vector()` function in `assertions.R`? `assert_that()` error messages are rather verbose...

File: R/derive_date_vars.R [hunk 4]
```diff
@@ -427,20 +439,35 @@ compute_dtf <- function(dtc, dt) {
...
-  case_when(
-    (!is_na & n_chr >= 19 & valid_dtc) | is_na | !valid_dtc ~ NA_character_,
-    n_chr == 16 ~ "S",
-    n_chr == 13 ~ "M",
-    n_chr == 10 | (n_chr > 0 & n_chr < 10) ~ "H"
-  )
+
+  if (ignore_seconds_flag)  {
+    if ((any(n_chr >= 17)))
+    {abort("Seconds detected in data while ignore_seconds_flag is invoked")
```
Comment (thomas-neitmann) [inline R/derive_date_vars.R:457]: The lints comment says it all: "Opening curly braces should never go on their own line and should always be followed by a new line."

File: R/derive_date_vars.R [hunk 5]
```diff
@@ -427,20 +439,35 @@ compute_dtf <- function(dtc, dt) {
...
+    }
+    else {
```
Comment (thomas-neitmann) [inline R/derive_date_vars.R:459]: Please put `else` on the same line as the closing parenthesis of `if`, i.e.

```
} else {
  ...
}
```

File: R/derive_date_vars.R [hunk 6]
```diff
@@ -427,20 +439,35 @@ compute_dtf <- function(dtc, dt) {
...
+      case_when(
+        (!is_na & n_chr >= 19 & valid_dtc) | is_na | !valid_dtc ~ NA_character_,
+        n_chr == 13 ~ "M",
+        n_chr == 10 | (n_chr > 0 & n_chr < 10) ~ "H"
+      )}
```
Comment (thomas-neitmann) [inline R/derive_date_vars.R:464]: lintr: "Closing curly-braces should always be on their own line, unless it's followed by an else."

File: R/derive_date_vars.R [hunk 7]
```diff
@@ -427,20 +439,35 @@ compute_dtf <- function(dtc, dt) {
...
+  } else {
+    case_when(
+      (!is_na & n_chr >= 19 & valid_dtc) | is_na | !valid_dtc ~ NA_character_,
+      n_chr == 16 ~ "S",
+      n_chr == 13 ~ "M",
+      n_chr == 10 | (n_chr > 0 & n_chr < 10) ~ "H")}
```
Comment (thomas-neitmann) [inline R/derive_date_vars.R:470]: lintr: "Closing curly-braces should always be on their own line, unless it's followed by an else."

File: R/derive_date_vars.R [hunk 8]
```diff
@@ -724,9 +777,15 @@ derive_vars_dtm <- function(dataset,
     # add --TMF variable
     tmf <- paste0(new_vars_prefix, "TMF")
     warn_if_vars_exist(dataset, tmf)
+
     dataset <- dataset %>%
-      mutate(!!sym(tmf) := compute_tmf(dtc = !!dtc, dtm = !!sym(dtm)))
+      mutate(!!sym(tmf) := compute_tmf(
+        dtc = !!dtc,
+        dtm = !!sym(dtm),
+        ignore_seconds_flag = ignore_seconds_flag))
```
Comment (thomas-neitmann) [inline R/derive_date_vars.R:785]: Closing parenthesis should be on its own line.

**Review by thomas-neitmann — CHANGES_REQUESTED** [review]

File: NEWS.md [hunk 9]
```diff
@@ -1,7 +1,12 @@
 # admiral development version
 
+- `derive_vars_dtm` now has `ignore_seconds_flag` argument so users can suppress `S` flag if seconds
```
Comment (bundfussr) [inline NEWS.md:3]: Please add brackets, i.e., use "`derive_vars_dtm()`"

**Review by bundfussr — CHANGES_REQUESTED** [review]

File: R/derive_date_vars.R (hunk 3 above)
Comment (bms63) [reply R/derive_date_vars.R:446]: Could this be split out into another issue - the `assert_date_vector()`?

Comment (bms63) [reply R/derive_date_vars.R:446]:
```
  assert_that(is_date(dtm))
  assert_character_vector(dtc)
  assert_logical_scalar(ignore_seconds_flag)
```
For now I just used the `assert_character_vector(dtc)` as you suggested

Comment (bms63) [reply R/derive_date_vars.R]: @thomas-neitmann  actually - is_date() performs this check already 
```
 assert_that(is_date("2019-07-18T15:25:00"))
Error: Argument "2019-07-18T15:25:00" = 2019-07-18T15:25:00 is not a lubridate date.
```

Comment (bms63) [reply R/derive_date_vars.R]: but maybe we should have a quick chat about it - if i do 

```
> is_date("2020-12-06")
[1] FALSE
```
then I don't get the Error message so I am happy to write a specific `assert_date_vector()` function, but can we move to a separate issue?  This `assert_that()` is used throughout this file so would need some more re-work.

**Review by bundfussr — APPROVED** [review]

File: R/derive_date_vars.R (hunk 3 above)
Comment (bms63) [reply R/derive_date_vars.R]: @thomas-neitmann Hey Thomas - can we make this into a separate issue.  It requires a bit of updating to the whole script and not just this section.

Comment (thomas-neitmann) [reply R/derive_date_vars.R]: Putting that into a spearate issue is fine with me 🙂

**Review by thomas-neitmann — APPROVED** [review]

**Pull request merged**